### ISSUE TITLE
Custom caption controls are not accessibility friendly

### ISSUE THREAD

**Issue by pananapread** [electron/electron#25984]
### Preflight Checklist

* [ ] I have read the [Contributing Guidelines](https://github.com/electron/electron/blob/master/CONTRIBUTING.md) for this project.
* [x] I agree to follow the [Code of Conduct](https://github.com/electron/electron/blob/master/CODE_OF_CONDUCT.md) that this project adheres to.
* [x] I have searched the issue tracker for an issue that matches the one I want to file, without success.

### Issue Details

* **Electron Version:**
  * 9.0.4
* **Operating System:**
  * Windows 10
* **Last Known Working Electron version:**
  * unknown

### Expected Behavior
Even with custom caption controls, on cursor hover Electron needs to respond to the WM_NCHITTEST window message with the appropriate parts. For example, Electron needs to respond to WM_NCHITTEST with HTMAXIMIZE for the maximize button so that we know when the mouse is over it

### Actual Behavior
Custom caption controls are not accessibility friendly because they are not properly labelled. For example, it doesn't respond to the WM_NCHITTEST window message with HTMAXIMIZE which leaves us with no way to reliable way to detect the maximize button. Apart from causing developer complications, this also causes some glaring accessibility issues that need to be dealt with. Users using a screen reader cannot identify custom caption controls.

### To Reproduce
This issue can be seen in any Electron app with custom caption controls (for example Microsoft Teams, Microsoft Office applications, Microsoft Edge, Visual Studio Code...)

The window control buttons are rendered using HTML, the window does not have a standard titlebar. The whole window is client area, that’s why it does not report HTMAXIMIZE, HTMINBUTTON or HTCLOSE in WM_NCHITTEST handler.

In the Electron code https://github.com/electron/electron/search?q=PreHandleMSG&unscoped_q=PreHandleMSG it looks like NativeWindowViews::PreHandleMSG gets called by Chromium for every window message.PreHandleMSG calls NotifyWindowMessage which is a multiplexer that notifies any listeners about the message but does not provide a mechanism to return a result. PreHandleMSG also does a bit of handling of some messages itself but does not handle WM_NCHITTEST. From that perspective, it looks like Electron would need to be updated to handle this case.
 
In the Chromium code: https://source.chromium.org/chromium/chromium/src/+/master:ui/views/win/hwnd_message_handler.cc;l=2246 it looks like it will call in to Electron via GetNonClientComponent(point) which would be another opportunity to return the correct HT value. It does not look like Electron handles that at all unfortunately.

### Screenshots
![image](https://user-images.githubusercontent.com/12780104/96189144-d26b7980-0ef4-11eb-9808-74cb790e79d2.png)
Logging is on the left. The Hello World! App uses the standard window titlebar which is why GetDwmAdornmentInfoRect returns success. It isn’t until apps start customizing their titlebar that we run into problems.

### Additional Information
Feel free to reach out to me with questions / if clarification is needed

**Comment by codebytere** [comment]
@apfoerts could you please provide some testable code that _does_ use custom caption controls?

**Comment by ckerr** [comment]
Would it be possible for you to make a standalone testcase with only the code necessary to reproduce the issue? For example, [Electron Fiddle](https://www.electronjs.org/fiddle) is great for quick little test cases and that Fiddle can publish to a [gist](https://gist.github.com) for testing by Electron maintainers

**Comment by TheMrJukes** [comment]
https://gist.github.com/TheMrJukes/5a7afc17cdb469715d1044348a97b207

Let me know if that works for you. The basic issue is that there is no way to specify that the html objects representing minimize, restore/maximize, and close map to their functionality so that they can be properly hit-tested by the platform.

**Comment by codebytere** [comment]
@TheMrJukes perfect! thanks :D

**Comment by apfoerts** [comment]
@ckerr what does it mean for this item to be moved from Backlog to Unsorted? I'm curious about what other information we can provide to make sure this work is prioritized :)

**Comment by ckerr** [comment]
@apfoerts, my apologies for missing your ping. I do think this is probably enough to start prioritizing & have moved the issue into that column.

**Comment by apfoerts** [comment]
Great! @ckerr @mlaurencin please let me know what other details we can provide here

**Comment by ckerr** [comment]
One of the next things is to figure out what API we should use for Electron to understand which custom elements map to these buttons.

* For example, the most heavyweight approach would be adding a totally new custom attribute. If we have to patch Chromium, then we have to maintain that patch. We should avoid custom attributes if possible. :slightly_smiling_face: 

* If we had an Electron-specific API that took in an object of role / element IDs so that Electron could look up the buttons by ID, that might be less invasive in terms of patches.

* Or if we were to reserve `data-` attribute names, e.g. `data-electron-wm-close` -- that feels a little clunky to me, but it would be doable.

@apfoerts do you have any opinions on what would be good for your work? It sounds like you're looking to get unblocked from _some_ solution rather than a _specific_ solution -- but even so, if you have thoughts on what would be preferable, that would be great to hear.

Also we should loop @electron/wg-api in to get their feedback.

**Comment by ckerr** [comment]
OK, a quick braindump before I EOD.

After a little testing, it looks like the PreHandleMSG() methods are never being reached. @apfoerts was correct with her second candidate -- all the action is happening in [HWNDMessageHandler::OnNCHitTest](https://source.chromium.org/chromium/chromium/src/+/master:ui/views/win/hwnd_message_handler.cc;l=2258;drc=db9ae7941adc1d95c943accce9e0151d265fd640). It's currently always returning `HTCLIENT` due to [this line](https://source.chromium.org/chromium/chromium/src/+/master:ui/views/win/hwnd_message_handler.cc;l=2275;drc=db9ae7941adc1d95c943accce9e0151d265fd640), and we can have it return a different value, e.g. [HTMAXBUTTON](https://source.chromium.org/chromium/chromium/src/+/master:ui/base/hit_test.h;l=29;drc=830286c24db2278d2169aceffdf27f6815a6c544), by overriding `HWNDMessageHandler::GetNonClientComponent()` in ElectronDesktopWindowTreeHostWin.

For example, I threw up a proof-of-concept that thinks the maximize button fills the entire width of the top 30 pixels:

```diff
diff --git a/shell/browser/ui/win/electron_desktop_window_tree_host_win.h b/shell/browser/ui/win/electron_desktop_window_tree_host_win.h
index ef9617547..e50f5919d 100644
--- a/shell/browser/ui/win/electron_desktop_window_tree_host_win.h
+++ b/shell/browser/ui/win/electron_desktop_window_tree_host_win.h
@@ -30,6 +30,7 @@ class ElectronDesktopWindowTreeHostWin
   bool GetDwmFrameInsetsInPixels(gfx::Insets* insets) const override;
   bool GetClientAreaInsets(gfx::Insets* insets,
                            HMONITOR monitor) const override;
+  int GetNonClientComponent(const gfx::Point& point) const override;

  private:
   NativeWindowViews* native_window_view_;  // weak ref
diff --git a/shell/browser/ui/win/electron_desktop_window_tree_host_win.cc b/shell/browser/ui/win/electron_desktop_window_tree_host_win.cc
index 852de00a5..2ce08a36c 100644
--- a/shell/browser/ui/win/electron_desktop_window_tree_host_win.cc
+++ b/shell/browser/ui/win/electron_desktop_window_tree_host_win.cc
@@ -99,4 +106,15 @@ bool ElectronDesktopWindowTreeHostWin::GetClientAreaInsets(
   return false;
 }

+int ElectronDesktopWindowTreeHostWin::GetNonClientComponent(
+    const gfx::Point& point) const {
+  if (point.y() < 31) {
+    // NOCOMMIT(ckerr) This is just a test for proof-of-concept
+    return HTMAXBUTTON;
+  }
+
+  return HTCLIENT;
+}
+
 }  // namespace electron
```

... and this value percolates back up to [HWNDMessageHandler::OnNCHitTest](https://source.chromium.org/chromium/chromium/src/+/master:ui/views/win/hwnd_message_handler.cc;l=2258;drc=db9ae7941adc1d95c943accce9e0151d265fd640) and appears to be returned to the system the way we'd want it to be. (NB: I have not confirmed this with an a11y tool).

So to sum up:

1. We need to test this with an a11y tool to confirm this approach works
1. We need some API (even if it's just a draft API) to have apps tell Electron when an HTML element is a min button, max button, close button, etc. 
1. We need to implement `ElectrionDesktopWindowTreeHostWin::GetNonClientComponent(const gfx::Point& point) const` to look up what HTML Element is at that x,y, use the information from the previous step to get its type, and return it here, defaulting to `HTCLIENT` if the user hasn't specified any value.
    1. Research needed on the cleanest way to to convert the x & y coordinate to an HTML Element.
1. We need tests :slightly_smiling_face:

**Comment by apfoerts** [comment]
Hi @ckerr, sorry for the delay in response (Hope you had a pleasant Thanksgiving 🦃) I was hoping to get some feedback from developers on implementation methods that would be preferable, but didn't get a chance to hear thoughts before we went on Holiday. I'll see if I can get some opinions!

Anything else I can help answer?

**Comment by mlaurencin** [comment]
I have some follow-up information for the items @ckerr  mentioned last week (bulleted numbers relate to the corresponding numbers in that post):

1. I tried implementing this change and used the screen reader NVDA to see if it would register that area as the maximize button. While `HTMAXBUTTON` was returned when that top 30 pixel area was scrolled over, the screen reader did not read it as a maximize button.
2. No further update
3. I looked into utilizing this [ElementFromPoint](https://source.chromium.org/chromium/chromium/src/+/master:third_party/blink/renderer/core/dom/document.h;drc=d81c5852498699fe3cd812e78d31c77c28e29281;l=410) function, but the `Document` object that calls it requires a reference to the HTML document the element is in. I am trying to determine if there is a nice way of getting this reference or if there is another potential option.
4. No further update

@apfoerts I do not believe we have any additional questions at the moment, but that feedback from developers would be appreciated if you can get it. (Also hope you had a good Thanksgiving as well 🙂)

**Comment by mlaurencin** [comment]
Hi @apfoerts before proceeding we just want to give an update and confirm what your desired behavior is.

At the moment, we are able to respond to `WM_NCHITTEST` with return values such as `HTMAXIMIZE` and `HTCLOSE`. The next step is to implement it so that the return value corresponds with the specified custom window control button. Is this your bar for success? And if so, do you have a tool/method of testing this on your end, and could you inform us of what that is so we can test on our end as well?

On a related note, to add onto my point `1` from last week, when that area was scrolled over it did not produce a tooltip saying `Maximize`. This differs from the system window control buttons. However, when `HTCLOSE` is returned instead, the tooltip `Close` does appear on mouseover. Would this functionality for all custom window control button be needed for success? And is there anything else, for example screen-reader functionality, needed for success for this issue?

**Comment by apfoerts** [comment]
Hi @mlaurencin, thank you for all the progress! I'm hoping to get some feedback from @TheMrJukes who's unfortunately OOF right now until the new year :/ 

The added info to #1 is interesting -- I'm not sure why it wouldn't be returning maximize and again I would have to wait for Bret's expertise here..

**Comment by TheMrJukes** [comment]
@mlaurencin Thanks for the updates!

I wrote a quick little tool that can help test: https://github.com/TheMrJukes/tools/tree/main/HitTester
It basically sends WM_NCHITTEST to the window under the cursor once every second and prints out the result.
Let me know if you need anything further there.

Am I understanding correctly that if you return HTMAXBUTTON then you do not get a "Maximize" (or "Restore" if the window is currently maximized) but the tooltips work correctly for HTMINBOX and HTCLOSE? That would be pretty unexpected. As far as I can tell from look at the code, all of those parts are treated roughly the same.

**Comment by mlaurencin** [comment]
@TheMrJukes Thank you for the tool! I will try to use it and let you know the results.

Regarding my previous update, you are correct regarding `HTMAXBUTTON` and `HTCLOSE`, but `HTMINBUTTON` (which I believe you were referring to by `HTMINBOX`, but please correct me if I am wrong) also does not work, in the same way as `HTMAXBUTTON`. I agree that it appears as though they should be treated the same, so if this is needed for success, we can look into it further. 

Additionally, I would like to confirm that the initial part of my previous message is what you all are looking for in terms of success. If there is some aspect that is missing or incorrect, please let us know.

**Comment by TheMrJukes** [comment]
Correct, I meant HTMINBUTTON. We need HTMINBUTTON, HTMAXBUTTON, and HTCLOSE to all be returned properly for their corresponding buttons/controls to consider success.

Our primary concern for this issue is responding to the WM_NCHITTEST message.
Screen readers generally use [MSAA ](https://docs.microsoft.com/en-us/windows/win32/winauto/microsoft-active-accessibility) to get more in depth information about an application/window. To my knowledge, there is nothing special about the caption control buttons wrt to MSAA. The screen reader will generally attempt to read the accessible name and role of the controls. It would be an accessibility win for Electron to have more MSAA support but that is tangential to this issue.

**Comment by mlaurencin** [comment]
Ok great, thank you for that clarification along with the information about MSAA @TheMrJukes.

I was able to test out your HitTester and we are able to respond with the `HTMINBUTTON`, `HTMAXBUTTON`, and `HTCLOSE` in our proof of concept setup. @ckerr and I will be pairing on this issue next week and working to have the corresponding buttons respond with their appropriate signal. We will let you know once we make any additional progress.

**Comment by apfoerts** [comment]
@mlaurencin I was wondering if there is any update here?

**Comment by mlaurencin** [comment]
Hi @apfoerts no real updates at the moment unfortunately. We are currently working towards the implementation mentioned in point 3 [above](https://github.com/electron/electron/issues/25984#issuecomment-732661304). Once we make some progress with this though we will definitely update again.

**Comment by apfoerts** [comment]
Happy March! 🍀 Just wanted to check in and see if there was anything we could help with :)

**Comment by mlaurencin** [comment]
Hi @apfoerts I believe there was another call sent out yesterday for feedback on what implementation developers would prefer (eg. HTML attribute, JavaScript function, etc.). Once we hear back regarding this, we should be in a better position to move forward. We will still update with any progress in the meantime however.
